### PR TITLE
Added guidelines on setting up screengrab tool and generating form-widgets guide

### LINKED ISSUE #486: Add guidelines on setting up screengrab tool and generating form-widgets guide
In reference to opendatakit/collect#1788, #428 .

### PR DESCRIPTION
**Pull request by Divya063**
closes #486 

## What is included in this PR?
Information on how to automate the generation of screenshots for form-widgets guide as well as regenerating form-widgets guide

### CODE REVIEW

**Comment by wonderchook** [comment]
thanks @Divya063 

@adammichaelwood do you think links to help sites is okay or should the instructions be directly in our documentation?

File: index.rst [hunk 1]
```diff
@@ -77,6 +77,7 @@ For a complete list of our projects, check out `Open Data Kit on Github <https:/
   form-update
   pyxform
   validate
+  screenshots
```
Comment (adammichaelwood) [inline index.rst]: I think this guide should probably be in the contributor's sections, not the form building section.

**Comment by adammichaelwood** [comment]
I don't have a problem with links to other help guides in theory -- we've done it several times.
It's sort of a judgement call on whether it makes it more clear or less clear.

I will spend some time on the guide and the tool today.

**Comment by wonderchook** [comment]
I was wondering if it made sense to link to Stackover flow. I can see how other guides make sense.

**Comment by adammichaelwood** [comment]
> link to Stackover flow.

I know we do this in at least one other place.
My personal feeling is that if the answer is old enough that it seems pretty mature and unlikely to change, it's fine.
But we haven't made any hard rules about it...

**Comment by adammichaelwood** [comment]
@Divya063 

I have (finally) taken some time to try to work through this and review. 
I'm not sure if what is here is usable in its current form, but it's hard for me to tell. I have tried a number of things, and I cannot get my local Ruby running things appropriately. This is not your fault; but using Ruby gems to get this done is presenting problems for me.

Since I can't actually run everything, I'm trying to just follow your instructions to get a sense of how everything works. I'm confused about having to git clone and build the app locally. It's not clear from your instructions why this is needed. Is it not possible to grab the screenshots from the live app on an Android device?

Since I can't actually complete the process here, I also can't comment on whether the output is what we're looking for.

Is there a way to simplify this process? I've spent several hours on it and haven't really been able to make headway. I'll ping you on Slack --- maybe we can schedule a call/demo soon.

File: screenshots.rst [hunk 2]
```diff
@@ -0,0 +1,263 @@
...
+
+.. code-block:: console
+
+  $ gem -v
+
+If it is not installed, you can install it using the instructions `given here <https://rubygems.org/pages/download>`_. Windows users are recommended to use `RubyInstallers <https://rubyinstaller.org/>`_.
+
+After installing RubyGem install the gems:
+
+Windows and Mac users can use the following commands:
```
Comment (yanokwa) [inline screenshots.rst]: On a stock macOS 10.13.3 running ruby 2.3, I get this error

```
mkmf.rb can't find header files for ruby at /System/Library/Frameworks/Ruby.framework/Versions/2.3/usr/lib/ruby/include/ruby.h
```

I fixed this by running `xcode-select --install`. 

Then I got another a permission error.
```
You don't have write permissions for the /Library/Ruby/Gems/2.3.0 directory
```

I fixed that by `sudo gem install fastlane`

This is roughly inline with what the fastlane docs say at https://docs.fastlane.tools/getting-started/ios/setup/.


@adammichaelwood Is this the error you got?

Comment (adammichaelwood) [reply screenshots.rst]: No. I got some other errors regarding the Ruby version (I was on 2.0, not 2.3).
Something, somewhere said I needed to be on 2.4, and several attempts to upgrade failed.

However --- I have successfully used Homebrew to upgrade to 2.3, and am trying your procedure above.....

Comment (adammichaelwood) [reply screenshots.rst]: I am now sorted. gems installed. i will run the screenshot script early tomorrow morning and report back

**Comment by adammichaelwood** [comment]
I got through the instructions as far as running building the debug APL

```
FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':collect_app:connectedDebugAndroidTest'.
> There were failing tests. See the report at: file:///Users/adamwood/odk/collect/collect-app/collect_app/build/reports/androidTests/connected/index.html

* Exception is:
org.gradle.api.tasks.TaskExecutionException: Execution failed for task ':collect_app:connectedDebugAndroidTest'.
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.executeActions(ExecuteActionsTaskExecuter.java:103)
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.execute(ExecuteActionsTaskExecuter.java:73)
        at org.gradle.api.internal.tasks.execution.OutputDirectoryCreatingTaskExecuter.execute(OutputDirectoryCreatingTaskExecuter.java:51)
        at org.gradle.api.internal.tasks.execution.SkipUpToDateTaskExecuter.execute(SkipUpToDateTaskExecuter.java:59)
        at org.gradle.api.internal.tasks.execution.ResolveTaskOutputCachingStateExecuter.execute(ResolveTaskOutputCachingStateExecuter.java:54)
        at org.gradle.api.internal.tasks.execution.ValidatingTaskExecuter.execute(ValidatingTaskExecuter.java:59)
        at org.gradle.api.internal.tasks.execution.SkipEmptySourceFilesTaskExecuter.execute(SkipEmptySourceFilesTaskExecuter.java:101)
        at org.gradle.api.internal.tasks.execution.FinalizeInputFilePropertiesTaskExecuter.execute(FinalizeInputFilePropertiesTaskExecuter.java:44)
        at org.gradle.api.internal.tasks.execution.CleanupStaleOutputsExecuter.execute(CleanupStaleOutputsExecuter.java:88)
        at org.gradle.api.internal.tasks.execution.ResolveTaskArtifactStateTaskExecuter.execute(ResolveTaskArtifactStateTaskExecuter.java:62)
        at org.gradle.api.internal.tasks.execution.SkipTaskWithNoActionsExecuter.execute(SkipTaskWithNoActionsExecuter.java:52)
        at org.gradle.api.internal.tasks.execution.SkipOnlyIfTaskExecuter.execute(SkipOnlyIfTaskExecuter.java:54)
        at org.gradle.api.internal.tasks.execution.ExecuteAtMostOnceTaskExecuter.execute(ExecuteAtMostOnceTaskExecuter.java:43)
        at org.gradle.api.internal.tasks.execution.CatchExceptionTaskExecuter.execute(CatchExceptionTaskExecuter.java:34)
        at org.gradle.execution.taskgraph.DefaultTaskGraphExecuter$EventFiringTaskWorker$1.run(DefaultTaskGraphExecuter.java:248)
        at org.gradle.internal.progress.DefaultBuildOperationExecutor$RunnableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:336)
        at org.gradle.internal.progress.DefaultBuildOperationExecutor$RunnableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:328)
        at org.gradle.internal.progress.DefaultBuildOperationExecutor.execute(DefaultBuildOperationExecutor.java:199)
        at org.gradle.internal.progress.DefaultBuildOperationExecutor.run(DefaultBuildOperationExecutor.java:110)
        at org.gradle.execution.taskgraph.DefaultTaskGraphExecuter$EventFiringTaskWorker.execute(DefaultTaskGraphExecuter.java:241)
        at org.gradle.execution.taskgraph.DefaultTaskGraphExecuter$EventFiringTaskWorker.execute(DefaultTaskGraphExecuter.java:230)
        at org.gradle.execution.taskgraph.DefaultTaskPlanExecutor$TaskExecutorWorker.processTask(DefaultTaskPlanExecutor.java:123)
        at org.gradle.execution.taskgraph.DefaultTaskPlanExecutor$TaskExecutorWorker.access$200(DefaultTaskPlanExecutor.java:79)
        at org.gradle.execution.taskgraph.DefaultTaskPlanExecutor$TaskExecutorWorker$1.execute(DefaultTaskPlanExecutor.java:104)
        at org.gradle.execution.taskgraph.DefaultTaskPlanExecutor$TaskExecutorWorker$1.execute(DefaultTaskPlanExecutor.java:98)
        at org.gradle.execution.taskgraph.DefaultTaskExecutionPlan.execute(DefaultTaskExecutionPlan.java:623)
        at org.gradle.execution.taskgraph.DefaultTaskExecutionPlan.executeWithTask(DefaultTaskExecutionPlan.java:578)
        at org.gradle.execution.taskgraph.DefaultTaskPlanExecutor$TaskExecutorWorker.run(DefaultTaskPlanExecutor.java:98)
        at org.gradle.internal.concurrent.ExecutorPolicy$CatchAndRecordFailures.onExecute(ExecutorPolicy.java:63)
        at org.gradle.internal.concurrent.ManagedExecutorImpl$1.run(ManagedExecutorImpl.java:46)
        at org.gradle.internal.concurrent.ThreadFactoryImpl$ManagedThreadRunnable.run(ThreadFactoryImpl.java:55)
Caused by: org.gradle.api.GradleException: There were failing tests. See the report at: file:///Users/adamwood/odk/collect/collect-app/collect_app/build/reports/androidTests/connected/index.html
        at com.android.build.gradle.internal.tasks.DeviceProviderInstrumentTestTask.runTests(DeviceProviderInstrumentTestTask.java:185)
        at org.gradle.internal.reflect.JavaMethod.invoke(JavaMethod.java:73)
        at org.gradle.api.internal.project.taskfactory.StandardTaskAction.doExecute(StandardTaskAction.java:46)
        at org.gradle.api.internal.project.taskfactory.StandardTaskAction.execute(StandardTaskAction.java:39)
        at org.gradle.api.internal.project.taskfactory.StandardTaskAction.execute(StandardTaskAction.java:26)
        at org.gradle.api.internal.AbstractTask$TaskActionWrapper.execute(AbstractTask.java:788)
        at org.gradle.api.internal.AbstractTask$TaskActionWrapper.execute(AbstractTask.java:755)
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter$1.run(ExecuteActionsTaskExecuter.java:124)
        at org.gradle.internal.progress.DefaultBuildOperationExecutor$RunnableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:336)
        at org.gradle.internal.progress.DefaultBuildOperationExecutor$RunnableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:328)
        at org.gradle.internal.progress.DefaultBuildOperationExecutor.execute(DefaultBuildOperationExecutor.java:199)
        at org.gradle.internal.progress.DefaultBuildOperationExecutor.run(DefaultBuildOperationExecutor.java:110)
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.executeAction(ExecuteActionsTaskExecuter.java:113)
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.executeActions(ExecuteActionsTaskExecuter.java:95)
        ... 30 more



```
Just to see, I tried moving forward anyway to using the screengrab tool.
That also (not surprising) errored out.

Here is the last part, starting with where I select which apk to use...

```
[07:45:07]: Select your debug app APK
1. collect_app/build/outputs/apk/androidTest/debug/collect-debug-androidTest-v1.13.0-3-ge0349138.apk
2. collect_app/build/outputs/apk/debug/collect-debug-v1.13.0-3-ge0349138.apk
3. collect_app/build/outputs/apk/odkCollectRelease/collect-odkCollectRelease-unsigned-v1.13.0-3-ge0349138.apk
4. collect_app/build/outputs/apk/release/collect-release-unsigned-v1.13.0-3-ge0349138.apk
?  2
[07:45:50]: To not be asked about this value, you can specify it using 'tests_apk_path'
[07:45:50]: Select your debug tests APK
1. collect_app/build/outputs/apk/androidTest/debug/collect-debug-androidTest-v1.13.0-3-ge0349138.apk
2. collect_app/build/outputs/apk/debug/collect-debug-v1.13.0-3-ge0349138.apk
3. collect_app/build/outputs/apk/odkCollectRelease/collect-odkCollectRelease-unsigned-v1.13.0-3-ge0349138.apk
4. collect_app/build/outputs/apk/release/collect-release-unsigned-v1.13.0-3-ge0349138.apk
?  1
[07:46:00]: The `aapt` command could not be found on your system, so your app APK could not be validated
[07:46:00]: Installing app APK
[07:46:00]: $ adb -s 04d919e10337c335 install -r collect_app/build/outputs/apk/debug/collect-debug-v1.13.0-3-ge0349138.apk
[07:46:02]: ▸ collect_app/build/outputs/apk/debug/collect-debug-v1.13.0-3-ge0349138.apk: 1 file pushed. 6.0 MB/s (9563994 bytes in 1.516s)
[07:46:02]: ▸ pkg: /data/local/tmp/collect-debug-v1.13.0-3-ge0349138.apk
[07:46:58]: ▸ Failure [INSTALL_FAILED_UPDATE_INCOMPATIBLE]

[!] App APK could not be installed
```




I have to go take care of kids for a few hours, I'll come back to this in the afternoon. Any advice in the meantime is helpful

**Comment by Divya063** [comment]
@adammichaelwood This error usually happens when the application is already installed

>  [07:46:00]: $ adb -s 04d919e10337c335 install -r collect_app/build/outputs/apk/debug/collect-debug-v1.13.0-3-ge0349138.apk
[07:46:02]: ▸ collect_app/build/outputs/apk/debug/collect-debug-v1.13.0-3-ge0349138.apk: 1 file pushed. 6.0 MB/s (9563994 bytes in 1.516s)
[07:46:02]: ▸ pkg: /data/local/tmp/collect-debug-v1.13.0-3-ge0349138.apk
[07:46:58]: ▸ Failure [INSTALL_FAILED_UPDATE_INCOMPATIBLE

You can run following commands to uninstall the existing version 
1. adb shell 
2. pm list packages
3. pm uninstall org.odk.collect.android

Or you can delete the test and debug apks  and try to generate new ones using `./gradlew assembleDebug assembleAndroidTest` 

For gradle build failure,  is it happening after running the command `./gradlew connectedAndroidTest` ?

**Comment by adammichaelwood** [comment]
> For gradle build failure, is it happening after running the command ./gradlew connectedAndroidTest ?

Yes

**Comment by adammichaelwood** [comment]
```
(odkenv)~/odk/collect/collect-app => ./gradlew connectedAndroidTest

> Configure project :collect_app 
Configuration 'androidTestCompile' in project ':collect_app' is deprecated. Use 'androidTestImplementation' instead.
Configuration 'compile' in project ':collect_app' is deprecated. Use 'implementation' instead.
registerResGeneratingTask is deprecated, use registerGeneratedFolders(FileCollection)
registerResGeneratingTask is deprecated, use registerGeneratedFolders(FileCollection)
registerResGeneratingTask is deprecated, use registerGeneratedFolders(FileCollection)

> Task :collect_app:processDebugGoogleServices 
Parsing json file: /Users/adamwood/odk/collect/collect-app/collect_app/src/debug/google-services.json

Unable to install /Users/adamwood/odk/collect/collect-app/collect_app/build/outputs/apk/debug/collect-debug-v1.13.0-3-ge0349138.apk
com.android.ddmlib.InstallException: INSTALL_FAILED_UPDATE_INCOMPATIBLE
        at com.android.ddmlib.Device.installRemotePackage(Device.java:1004)
        at com.android.ddmlib.Device.installPackage(Device.java:911)
        at com.android.builder.testing.ConnectedDevice.installPackage(ConnectedDevice.java:122)
        at com.android.builder.internal.testing.SimpleTestCallable.call(SimpleTestCallable.java:126)
        at com.android.builder.internal.testing.SimpleTestCallable.call(SimpleTestCallable.java:52)
        at java.util.concurrent.ForkJoinTask$AdaptedCallable.exec(ForkJoinTask.java:1424)
        at java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:289)
        at java.util.concurrent.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1056)
        at java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1692)
        at java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:157)

com.android.builder.testing.ConnectedDevice > runTests[Nexus 5 - 6.0.1] FAILED 
        com.android.builder.testing.api.DeviceException: com.android.ddmlib.InstallException: INSTALL_FAILED_UPDATE_INCOMPATIBLE
                at com.android.builder.testing.ConnectedDevice.installPackage(ConnectedDevice.java:128)

> Task :collect_app:connectedDebugAndroidTest 
[no message defined]
java.lang.RuntimeException: com.android.builder.testing.api.DeviceException: com.android.ddmlib.InstallException: INSTALL_FAILED_UPDATE_INCOMPATIBLE
        at java.util.concurrent.ForkJoinTask$AdaptedCallable.exec(ForkJoinTask.java:1431)
        at java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:289)
        at java.util.concurrent.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1056)
        at java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1692)
        at java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:157)
Caused by: com.android.builder.testing.api.DeviceException: com.android.ddmlib.InstallException: INSTALL_FAILED_UPDATE_INCOMPATIBLE
        at com.android.builder.testing.ConnectedDevice.installPackage(ConnectedDevice.java:128)
        at com.android.builder.internal.testing.SimpleTestCallable.call(SimpleTestCallable.java:126)
        at com.android.builder.internal.testing.SimpleTestCallable.call(SimpleTestCallable.java:52)
        at java.util.concurrent.ForkJoinTask$AdaptedCallable.exec(ForkJoinTask.java:1424)
        ... 4 more
Caused by: com.android.ddmlib.InstallException: INSTALL_FAILED_UPDATE_INCOMPATIBLE
        at com.android.ddmlib.Device.installRemotePackage(Device.java:1004)
        at com.android.ddmlib.Device.installPackage(Device.java:911)
        at com.android.builder.testing.ConnectedDevice.installPackage(ConnectedDevice.java:122)
        ... 7 more


FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':collect_app:connectedDebugAndroidTest'.
> There were failing tests. See the report at: file:///Users/adamwood/odk/collect/collect-app/collect_app/build/reports/androidTests/connected/index.html

* Try:
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output. Run with --scan to get full insights.

* Get more help at https://help.gradle.org

BUILD FAILED in 11s
57 actionable tasks: 4 executed, 53 up-to-date
```

**Comment by adammichaelwood** [comment]
Some Googling reveals this is part of a larger pattern of problems...
I tried:

```
./gradlew connectedCheck
```

And got a similar bunch of errors.

```
> Task :collect_app:connectedDebugAndroidTest FAILED
Unable to install /Users/adamwood/odk/collect/collect-app/collect_app/build/outputs/apk/debug/collect-debug-v1.13.0-3-ge0349138.apk
com.android.ddmlib.InstallException: INSTALL_FAILED_UPDATE_INCOMPATIBLE
```

**Comment by wonderchook** [comment]
@yanokwa @adammichaelwood are we good on this now? What changes need to be made?

File: src/screenshots.rst [hunk 3]
```diff
@@ -0,0 +1,275 @@
...
+
+If it is not installed, you can install it using the instructions `given here <https://rubygems.org/pages/download>`_. Windows users are recommended to use `RubyInstallers <https://rubyinstaller.org/>`_.
+
+After installing RubyGem install the gems:
+
+Windows and Mac users can use the following commands:
+
+.. code-block:: console
+
+  $ gem install fastlane
```
Comment (yanokwa) [inline src/screenshots.rst]: On Mac, you'll need to do a `xcode-select --install` first. Then `sudo` gem install fastlane`.

**Comment by yanokwa** [comment]
I've added my comments about the need for installing XCode's CLI tools first. @adammichaelwood what else did you have to do to get this working?

**Comment by adammichaelwood** [comment]
I also had to upgrade Ruby to 2.3.

```
brew install ruby
```

(There are a lot of instructions all over for upgrading Ruby. But, for whatever reason, that was the only one that worked for me. Even `brew upgrade ruby` didn't work for some reason, even though I had Ruby already. This might cause people more problems, since you have to also get homebrew if you don't have it, given the limited scope of users, that's probably ok.)

**Comment by yanokwa** [comment]
Macs often come with an old version of ruby, so we could recommend that people install a recent version of ruby, preferably through brew and with rbenv. https://github.com/rbenv/rbenv#homebrew-on-macos has rough instructions...

```
brew install rbenv
rbenv init
rbenv install 2.5.0
cd /path/to/collect
rbenv local 2.5.0
gem install fastlane
```

**Comment by Divya063** [comment]
I have made the required changes!

**Comment by yanokwa** [comment]
Closing because it's unlikely we'll do this in the foreseeable future.

**Pull request closed without merging**